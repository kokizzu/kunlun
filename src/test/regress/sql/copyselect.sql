--
-- Test cases for COPY (select) TO
--
--DDL_STATEMENT_BEGIN--
create table test1 (id serial, t text);
--DDL_STATEMENT_END--
insert into test1 (t) values ('a');
insert into test1 (t) values ('b');
insert into test1 (t) values ('c');
insert into test1 (t) values ('d');
insert into test1 (t) values ('e');

--DDL_STATEMENT_BEGIN--
create table test2 (id serial, t text);
--DDL_STATEMENT_END--
insert into test2 (t) values ('A');
insert into test2 (t) values ('B');
insert into test2 (t) values ('C');
insert into test2 (t) values ('D');
insert into test2 (t) values ('E');

--DDL_STATEMENT_BEGIN--
create view v_test1
as select 'v_'||t from test1;
--DDL_STATEMENT_END--

--
-- Test COPY table TO
--
copy test1 to stdout;
--
-- This should fail
--
copy v_test1 to stdout;
--
-- Test COPY (select) TO
--
copy (select t from test1 where id=1) to stdout;
--
-- Test COPY (select for update) TO
--
copy (select t from test1 where id=3 for update) to stdout;
--
-- This should fail
--
copy (select t into temp test3 from test1 where id=3) to stdout;
--
-- This should fail
--
copy (select * from test1) from stdin;
--
-- This should fail
--
copy (select * from test1) (t,id) to stdout;
--
-- Test JOIN
--
copy (select * from test1 join test2 using (id)) to stdout;
--
-- Test UNION SELECT
--
copy (select t from test1 where id = 1 UNION select * from v_test1 ORDER BY 1) to stdout;
--
-- Test subselect
--
copy (select * from (select t from test1 where id = 1 UNION select * from v_test1 ORDER BY 1) t1) to stdout;
--
-- Test headers, CSV and quotes
--
copy (select t from test1 where id = 1) to stdout csv header force quote t;
--
-- Test psql builtins, plain table
--
\copy test1 to stdout
--
-- This should fail
--
\copy v_test1 to stdout
--
-- Test \copy (select ...)
--
\copy (select "id",'id','id""'||t,(id + 1)*id,t,"test1"."t" from test1 where id=3) to stdout
--
-- Drop everything
--
--DDL_STATEMENT_BEGIN--
drop table test2;
--DDL_STATEMENT_END--
--DDL_STATEMENT_BEGIN--
drop view v_test1;
--DDL_STATEMENT_END--
--DDL_STATEMENT_BEGIN--
drop table test1;
--DDL_STATEMENT_END--

-- psql handling of COPY in multi-command strings
copy (select 1) to stdout\; select 1/0;	-- row, then error
select 1/0\; copy (select 1) to stdout; -- error only
copy (select 1) to stdout\; copy (select 2) to stdout\; select 0\; select 3; -- 1 2 3

--DDL_STATEMENT_BEGIN--
create table test3 (c int);
--DDL_STATEMENT_END--
select 0\; copy test3 from stdin\; copy test3 from stdin\; select 1; -- 1
1
\.
2
\.
select * from test3;
--DDL_STATEMENT_BEGIN--
drop table test3;
--DDL_STATEMENT_END--
