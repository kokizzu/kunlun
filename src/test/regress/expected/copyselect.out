--
-- Test cases for COPY (select) TO
--
create table test1 (id serial, t text);
CREATE TABLE
insert into test1 (t) values ('a');
INSERT 0 1
insert into test1 (t) values ('b');
INSERT 0 1
insert into test1 (t) values ('c');
INSERT 0 1
insert into test1 (t) values ('d');
INSERT 0 1
insert into test1 (t) values ('e');
INSERT 0 1
create table test2 (id serial, t text);
CREATE TABLE
insert into test2 (t) values ('A');
INSERT 0 1
insert into test2 (t) values ('B');
INSERT 0 1
insert into test2 (t) values ('C');
INSERT 0 1
insert into test2 (t) values ('D');
INSERT 0 1
insert into test2 (t) values ('E');
INSERT 0 1
create view v_test1
as select 'v_'||t from test1;
CREATE VIEW
--
-- Test COPY table TO
--
copy test1 to stdout;
1	a
2	b
3	c
4	d
5	e
--
-- This should fail
--
copy v_test1 to stdout;
psql:sql/copyselect.sql:28: ERROR:  cannot copy from view "v_test1"
HINT:  Try the COPY (SELECT ...) TO variant.
--
-- Test COPY (select) TO
--
copy (select t from test1 where id=1) to stdout;
a
--
-- Test COPY (select for update) TO
--
copy (select t from test1 where id=3 for update) to stdout;
c
--
-- This should fail
--
copy (select t into temp test3 from test1 where id=3) to stdout;
psql:sql/copyselect.sql:40: ERROR:  COPY (SELECT INTO) is not supported
--
-- This should fail
--
copy (select * from test1) from stdin;
psql:sql/copyselect.sql:44: ERROR:  syntax error at or near "from"
LINE 1: copy (select * from test1) from stdin;
                                   ^
--
-- This should fail
--
copy (select * from test1) (t,id) to stdout;
psql:sql/copyselect.sql:48: ERROR:  syntax error at or near "("
LINE 1: copy (select * from test1) (t,id) to stdout;
                                   ^
--
-- Test JOIN
--
copy (select * from test1 join test2 using (id)) to stdout;
1	a	A
2	b	B
3	c	C
4	d	D
5	e	E
--
-- Test UNION SELECT
--
copy (select t from test1 where id = 1 UNION select * from v_test1 ORDER BY 1) to stdout;
a
v_a
v_b
v_c
v_d
v_e
--
-- Test subselect
--
copy (select * from (select t from test1 where id = 1 UNION select * from v_test1 ORDER BY 1) t1) to stdout;
a
v_a
v_b
v_c
v_d
v_e
--
-- Test headers, CSV and quotes
--
copy (select t from test1 where id = 1) to stdout csv header force quote t;
t
"a"
--
-- Test psql builtins, plain table
--
\copy test1 to stdout
1	a
2	b
3	c
4	d
5	e
--
-- This should fail
--
\copy v_test1 to stdout
psql:sql/copyselect.sql:72: ERROR:  cannot copy from view "v_test1"
HINT:  Try the COPY (SELECT ...) TO variant.
--
-- Test \copy (select ...)
--
\copy (select "id",'id','id""'||t,(id + 1)*id,t,"test1"."t" from test1 where id=3) to stdout
3	id	id""c	12	c	c
--
-- Drop everything
--
drop table test2;
DROP TABLE
drop view v_test1;
DROP VIEW
drop table test1;
DROP TABLE
-- psql handling of COPY in multi-command strings
copy (select 1) to stdout\; select 1/0;	-- row, then error
1
psql:sql/copyselect.sql:85: ERROR:  division by zero
select 1/0\; copy (select 1) to stdout; -- error only
psql:sql/copyselect.sql:86: ERROR:  division by zero
copy (select 1) to stdout\; copy (select 2) to stdout\; select 0\; select 3; -- 1 2 3
1
2
 ?column? 
----------
        3
(1 row)

create table test3 (c int);
CREATE TABLE
select 0\; copy test3 from stdin\; copy test3 from stdin\; select 1; -- 1
 ?column? 
----------
        1
(1 row)

select * from test3;
 c 
---
 1
 2
(2 rows)

drop table test3;
DROP TABLE
